libs/community/langchain_community/vectorstores/cratedb/base.py

```diff
@@ -0,0 +1,453 @@
+                raise ValueError("Collection not found")
+            return self._query_collection_multi(
+                collections=[collection], embedding=embedding, k=k, filter=filter
+            )
+
+    def _query_collection_multi(
+        self,
+        collections: List[Any],
+        embedding: List[float],
+        k: int = 4,
+        filter: Optional[Dict[str, str]] = None,
+    ) -> List[Any]:
+        """Query the collection."""
+        self._init_models(embedding)
+
+        collection_names = [coll.name for coll in collections]
+        collection_uuids = [coll.uuid for coll in collections]
+        self.logger.info(f"Querying collections: {collection_names}")
+
+        with self.Session() as session:
+            filter_by = self.EmbeddingStore.collection_id.in_(collection_uuids)
+
+            if filter is not None:
+                filter_clauses = []
+                for key, value in filter.items():
+                    IN = "in"
+                    if isinstance(value, dict) and IN in map(str.lower, value):
+                        value_case_insensitive = {
+                            k.lower(): v for k, v in value.items()
+                        }
+                        filter_by_metadata = self.EmbeddingStore.cmetadata[key].in_(
+                            value_case_insensitive[IN]
+                        )
+                        filter_clauses.append(filter_by_metadata)
+                    else:
+                        filter_by_metadata = self.EmbeddingStore.cmetadata[key] == str(
+                            value
+                        )  # type: ignore[assignment]
+                        filter_clauses.append(filter_by_metadata)
+
+                filter_by = sqlalchemy.and_(filter_by, *filter_clauses)  # type: ignore[assignment]
+
+            _type = self.EmbeddingStore
+
+            results: List[Any] = (
+                session.query(  # type: ignore[attr-defined]
+                    self.EmbeddingStore,
+                    # TODO: Original pgvector code uses `self.distance_strategy`.
+                    #       CrateDB currently only supports EUCLIDEAN.
+                    #       self.distance_strategy(embedding).label("distance")
+                    sqlalchemy.literal_column(
+                        f"{self.EmbeddingStore.__tablename__}._score"
+                    ).label("_score"),
+                )
+                .filter(filter_by)
+                # CrateDB applies `KNN_MATCH` within the `WHERE` clause.
+                .filter(
+                    sqlalchemy.func.knn_match(
+                        self.EmbeddingStore.embedding, embedding, k
+                    )
+                )
+                .order_by(sqlalchemy.desc("_score"))
+                .join(
+                    self.CollectionStore,
+                    self.EmbeddingStore.collection_id == self.CollectionStore.uuid,
+                )
+                .limit(k)
+            )
```

Comment:
I guess it is just NOT ok to use `_score` here, and this has just been applied in the interim?
Isn't this the spot where that feature request was coming from, to be able to use the actual vector distance?

- https://github.com/crate/crate/issues/15835

/cc @ckurze

Reply:
I dearly hope so!

Reply:
In the best case, just using a corresponding SQLAlchemy incantation (correctly) will work without much ado.

```python
import sqlalchemy as sa

sa.func.vector_similarity(...)
```

Reply:
Swapping in.
```python
sqlalchemy.func.vector_similarity(
    self.EmbeddingStore.embedding, embedding
).label("_score"),
```

Not there yet.
- https://github.com/crate/crate/issues/16912

Reply:
Fixed per 476d718b998e, with improvements pending for a later iteration.

Reply:
I'm sorry but I don't understand the problem, what is it that we want? vector distance? if so, we do no directly have it, our `vector_similarity` gives us similarity based on euclidian distance, similarity is calculated like `sim = 1 / distance`

Reply:
As per [helper function above](https://github.com/langchain-ai/langchain/pull/27710#discussion_r1821498985), I think `similarity / math.sqrt(2)` will be used for normalization.

Reply:
Relates to https://github.com/langchain-ai/langchain/pull/27710/#discussion_r1832222614, +1 to `vector_similarity` and there is no need to normalize it. 

If https://github.com/crate/crate/issues/16912 was a blocker, 5.9.3 will have it fixed.

Reply:
It's not a blocker, we have been able to work around it. Thanks.